pennylane/devices/qubit/simulate.py

```diff
@@ -335,6 +368,353 @@ def simulate_partial(k):
+        mcm_samples (dict): Mid-circuit measurement samples or all parent circuits of ``circuit``
+
+    Returns:
+        tuple(TensorLike): The results of the simulation
+    """
+    interface = execution_kwargs.get("interface", None)
+    postselect_mode = execution_kwargs.get("postselect_mode", None)
+
+    samples_present = any(isinstance(mp, SampleMP) for mp in circuit.measurements)
+    postselect_present = any(
+        op.postselect is not None for op in circuit.operations if isinstance(op, MidMeasureMP)
+    )
+    if postselect_present and samples_present and circuit.batch_size is not None:
+        raise ValueError(
+            "Returning qml.sample is not supported when postselecting mid-circuit "
+            "measurements with broadcasting"
+        )
+
+    #########################
+    # shot vector treatment #
+    #########################
+    if circuit.shots.has_partitioned_shots:
+        results = []
+        for s in circuit.shots:
+            aux_circuit = qml.tape.QuantumScript(
+                circuit.operations,
+                circuit.measurements,
+                shots=qml.measurements.Shots(s),
+                trainable_params=circuit.trainable_params,
+            )
+            results.append(simulate_tree_mcm(aux_circuit, debugger, **execution_kwargs))
+        return tuple(results)
+
+    #######################
+    # main implementation #
+    #######################
+
+    def init_dict(d):
+        return {} if d is None else d
+
+    mcm_active = init_dict(mcm_active)
+    mcm_samples = init_dict(mcm_samples)
+
+    circuit_base, circuit_next, op = circuit_up_to_first_mcm(circuit)
+    # we need to make sure the state is the all-wire state
+    initial_state = prep_initial_state(circuit_base, interface, initial_state, circuit.wires)
+    state, is_state_batched = get_final_state(
+        circuit_base,
+        debugger=debugger,
+        initial_state=initial_state,
+        mid_measurements=mcm_active,
+        **execution_kwargs,
+    )
+    measurements = measure_final_state(
+        circuit_base, state, is_state_batched, initial_state=initial_state, **execution_kwargs
+    )
+
+    # Simply return measurements when ``circuit_base`` does not have an MCM
+    if circuit_next is None:
+        return measurements
+
+    # For 1-shot measurements as 1-D arrays
+    if op.postselect is not None and postselect_mode == "fill-shots":
+        samples = op.postselect * qml.math.ones_like(measurements)
+    else:
+        samples = qml.math.atleast_1d(measurements)
+    update_mcm_samples(op, samples, mcm_active, mcm_samples)
+
+    # Define ``branch_measurement`` here to capture ``op``, ``rng``, ``prng_key``, ``debugger``, ``interface``
+    def branch_measurement(circuit_next, state, branch, mcm_active, mcm_samples):
+        """Returns the measurements of the specified branch by executing ``circuit_next``."""
```

Comment:
For these inner functions, can we explicitly document what closure variables are being used? This will greatly help with readability.